### ISSUE TITLE
How to get "Old Values" from AutoSavePreaction

### ISSUE THREAD

**Issue by ChuckBryan** [zzzprojects/EntityFramework-Plus#423]
Hi!

I am continuing work on the solution from this question: https://github.com/zzzprojects/EntityFramework-Plus/issues/404#issuecomment-419656560

What I noticed was that often, I would have the related entity, but, there were times when it was null. In that case, I used the FK value to look up the Description and then added that as the new Property Name Value.

However, I am stumped on how to get the "Old" Value? I was hoping that I could get what the FK Value was before the change. I've looked through the objects in Quick View, but don't see how to retrieve it.

To summarize - I can get the current values in AutoSavePreAction with the Audit.Entries. Is there a way to get what the entity was prior to the save?

Thanks

Chuck

**Comment by JonathanMagnan** [comment]
Hello @ChuckBryan ,

Do you think you could provide a project sample with this issue?

It will help my developer investigate the issue more efficiently.

_We now always ask for a project sample even if the issue is easy to reproduce. As a free product, we must find some way to save time to offer an overall better experience for everyone and release fixes faster._

_You can send it to: info@zzzprojects.com if you need to keep the source private_

Best Regards,

Jonathan

---

**Performance Libraries**
`context.BulkInsert(list, options => options.BatchSize = 1000);  `
[Entity Framework Extensions](http://entityframework-extensions.net/) • [Entity Framework Classic](http://entityframework-classic.net/) • [Bulk Operations](http://bulk-operations.net/) •  [Dapper Plus](http://dapper-plus.net/) •  [LinqToSql Plus](http://linqtosql-plus.net/)

**Runtime Evaluation**
`Eval.Execute("x + y", new {x = 1, y = 2}); // return 3`
[C# Eval Function](http://eval-expression.net/) • [SQL Eval Function](http://eval-sql.net/)

**Comment by ChuckBryan** [comment]
I am going to close this out. I understand where you are coming from with regards to a project sample to repo, but, I didn't think that my question is an issue, but rather a question. Perhaps I phrased it incorrectly. I had searched to see how to get the "Old Value" prior to the change so that I could do additional work, but couldn't find any.

**Comment by ChuckBryan** [comment]
Using the code below, I can create a custom audit. When I look in the "NewValue" column of the AuditEntryProperties, it has the expected value. However, the "OldValue" is null. How do I populate the Old Value?
```
using System.Data.SqlClient;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Z.Test.EntityFramework.Plus;

namespace Z.EntityFramework.Plus.Lab.EFCore20
{
    public static class Request_Audit_External
    {
        public static void Execute()
        {
            // CLEAR
            using (var ctx = new EntityContext())
            {
                ctx.Addresses.RemoveRange(ctx.Addresses);
                ctx.AddressTypes.RemoveRange(ctx.AddressTypes);
                ctx.AuditEntries.RemoveRange(ctx.AuditEntries);
                ctx.AuditEntryProperties.RemoveRange(ctx.AuditEntryProperties);
                ctx.SaveChanges();
            }

            // SEED
            using (var ctx = new EntityContext())
            {
                var addressType = ctx.AddressTypes.Add(new AddressType {Description = "Type1"}).Entity;
                ctx.Addresses.Add(new Address {AddressType = addressType, Street = "ZZZ Street"});

                ctx.SaveChanges();
            }

            // TEST
            using (var ctx = new EntityContext())
            {
                var addresses = ctx.Addresses.Include(x => x.AddressType).ToList();

                addresses[0].Street = "YYY Street";

                var audit = new Audit();
                ctx.SaveChanges(audit);
            }
        }

        public class EntityContext : DbContext
        {
            static EntityContext()
            {
                AuditManager.DefaultConfiguration.AutoSavePreAction = (context, audit) =>
                {
                    audit.Entries.ForEach(x =>
                    {
                        if (x.Entity is Address address && address.AddressType != null)
                        {
                            x.Properties.Add(new AuditEntryProperty
                            {
                                PropertyName = nameof(AddressType.Description),
                                NewValue = address.AddressType.Description
                            });
                        }
                    });
                    (context as EntityContext).AuditEntries.AddRange(audit.Entries);
                };

                var context2 = new EntityContext();
                context2.Database.EnsureCreated();
            }

            public DbSet<Address> Addresses { get; set; }
            public DbSet<AddressType> AddressTypes { get; set; }
            public DbSet<AuditEntry> AuditEntries { get; set; }
            public DbSet<AuditEntryProperty> AuditEntryProperties { get; set; }

            protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
            {
                optionsBuilder.UseSqlServer(new SqlConnection(My.Config.ConnectionStrings.TestDatabase));

                base.OnConfiguring(optionsBuilder);
            }
        }

        public class Address
        {
            public int ID { get; set; }
            public AddressType AddressType { get; set; }
            public string Street { get; set; }
        }

        public class AddressType
        {
            public int ID { get; set; }
            public string Description { get; set; }
        }
    }
}
```

**Comment by ChuckBryan** [comment]
I just realized that I am going to have to get the Old and New Value of the actual property that changed (that is, the Address Type Id) and then resolve the Description for both. Then, I can add that to the OldValue of the AuditEntryProperty.

I was able to find the Address Type Id for Old/New from the "x.Entity.Properties" and then filtering that collection by the Internal Property Name.

for example:

```
audit.Entries.ForEach(x =>
{
	if (x.Entity is Address address && address.AddressType != null)
	{

		var addressType = x.Properties.FirstOrDefault(z => z.InternalPropertyName == "AddressTypeId");

		if (addressType != null)
		{
			var oldAddressType  = ((ApplicationDbContext) context).AddressTypes.Find(addressType .OldValue);
			var newAddressType  = ((ApplicationDbContext) context).AddressTypes.Find(addressType .NewValue);


			x.Properties.Add(new AuditEntryProperty()
			{
				PropertyName = "AddressType",
				NewValue = newAddressType.Description,
				OldValue = oldAddressType.Description,
			});
		}
	}
});
```
